x-pack/plugins/lens/public/app_plugin/save_modal_container.tsx

```diff
@@ -0,0 +1,396 @@
+  }, [initLastKnowDoc]);
+
+  useEffect(() => {
+    if (initialInput && isVisible) {
+      async function loadLastKnownDoc() {
+        getLastKnownDoc({
+          data,
+          initialInput,
+          chrome,
+          notifications,
+          attributeService,
+        }).then((result) => {
+          if (result) setLastKnownDoc(result.doc);
+        });
+      }
+
+      loadLastKnownDoc();
+    }
+  }, [chrome, data, initialInput, notifications, attributeService, isVisible]);
+
+  const tagsIds =
+    persistedDoc && savedObjectsTagging
+      ? savedObjectsTagging.ui.getTagIdsFromReferences(persistedDoc.references)
+      : [];
+
+  const runLensSave = (saveProps: SaveProps, options: { saveToLibrary: boolean }) => {
+    if (runSave) {
+      // inside lens, we use the function that's passed to it
+      runSave(saveProps, options);
+    } else {
+      if (attributeService) {
+        runSaveLensVisualization(
+          {
+            ...lensServices,
+            lastKnownDoc,
+            setIsSaveModalVisible: () => {
+              onClose();
+            },
+            initialInput,
+            attributeService,
+            redirectTo,
+            redirectToDashboard,
+            redirectToOrigin,
+            getIsByValueMode: () => false,
+            onAppLeave: () => {},
+          },
+          saveProps,
+          options
+        ).then(() => {
+          onSave?.();
+          onClose();
+        });
+      }
+    }
+  };
+
+  const savingToLibraryPermitted = Boolean(isSaveable && application.capabilities.visualize.save);
+
+  return (
+    <SaveModal
+      isVisible={isVisible}
+      originatingApp={originatingApp}
+      savingToLibraryPermitted={savingToLibraryPermitted}
+      allowByValueEmbeddables={dashboardFeatureFlag?.allowByValueEmbeddables}
+      savedObjectsTagging={savedObjectsTagging}
+      tagsIds={tagsIds}
+      onSave={(saveProps, options) => {
+        runLensSave(saveProps, options);
+      }}
+      onClose={onClose}
+      getAppNameFromId={getAppNameFromId}
+      lastKnownDoc={lastKnownDoc}
+      returnToOriginSwitchLabel={returnToOriginSwitchLabel}
+    />
+  );
+}
+
+export const runSaveLensVisualization = async (
+  props: {
+    lastKnownDoc?: Document;
+    getIsByValueMode: () => boolean;
+    persistedDoc?: Document;
+    setIsSaveModalVisible: () => void;
+  } & ExtraProps &
+    LensAppServices,
+  saveProps: SaveProps,
+  options: { saveToLibrary: boolean }
+) => {
+  return new Promise<Partial<LensAppState>>(async (resolve, reject) => {
+    const {
+      chrome,
+      initialInput,
+      lastKnownDoc,
+      persistedDoc,
+      savedObjectsClient,
+      overlays,
+      notifications,
+      stateTransfer,
+      attributeService,
+      savedObjectsTagging,
+      redirectToDashboard,
+      getIsByValueMode,
+      redirectToOrigin,
+      onAppLeave,
+      setIsSaveModalVisible,
+      redirectTo,
+    } = props;
+    if (!lastKnownDoc) {
+      return;
```

Comment:
I am a bit concerned with all these `return` within the `Promise` constructor: is this extra wrapping really required?

Reply:
makes sense, i didn't want to change lot's of code, this is why it looks like the way it is.
Let me spend some time refactoring both